src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/AndroidBrokerProxy.cs

```diff
@@ -206,48 +151,166 @@ public AdalResultWrapper GetAuthTokenInBackground(AuthenticationRequest request,
+        {
+            Intent intent = null;
+            IAccountManagerFuture result = null;
+            try
+            {
+                // Callback is not passed since it is making a blocking call to get
+                // intent. Activity needs to be launched from calling app
+                // to get the calling app's metadata if needed at BrokerActivity.
+                Bundle addAccountOptions = GetBrokerOptions(request);
+                result = _androidAccountManager.AddAccount(BrokerConstants.BrokerAccountType,
+                    BrokerConstants.AuthtokenType, null, addAccountOptions, null,
+                    null, new Handler(callerActivity.MainLooper));
+
+                // Making blocking request here
+                Bundle bundleResult = (Bundle)result.Result;
+                // Authenticator should throw OperationCanceledException if
+                // token is not available
+                intent = (Intent)bundleResult.GetParcelable(AccountManager.KeyIntent);
+
+                // Add flag to this intent to signal that request is for broker
+                // logic
+                if (intent != null)
+                {
+                    intent.PutExtra(BrokerConstants.BrokerRequest, BrokerConstants.BrokerRequest);
+                }
+            }
+            catch (AdalException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                _logger.ErrorPii(e);
+            }
+
+            return intent;
+        }
+
+        // App needs to give permission to AccountManager to use broker.
+        private bool VerifyManifestPermissions()
+        {
+            return VerifyManifestPermission("android.permission.GET_ACCOUNTS") &&
+                   VerifyManifestPermission("android.permission.MANAGE_ACCOUNTS") &&
+                   VerifyManifestPermission("android.permission.USE_CREDENTIALS");
+        }
+
+        private bool VerifyManifestPermission(string permission)
+        {
+            if (Permission.Granted !=
+                Application.Context.PackageManager.CheckPermission(permission, Application.Context.PackageName))
+            {
+                _logger.Info(string.Format(CultureInfo.InvariantCulture,
+                    AdalErrorMessageAndroidEx.MissingPackagePermissionTemplate, permission));
+
+                return false;
+            }
+            return true;
+        }
+
+        private void VerifyNotOnMainThread()
+        {
+            Looper looper = Looper.MyLooper();
+            if (looper != null && looper == _androidContext.MainLooper)
+            {
+                Exception exception = new AdalException(
+                    "Calling this from your main thread can lead to deadlock");
+                _logger.ErrorPii(exception);
+
+                if (_androidContext.ApplicationInfo.TargetSdkVersion >= BuildVersionCodes.Froyo)
```

Comment:
>Froyo [](start = 90, length = 5)

what's this?

Reply:
:joy: 

---
In reply to: [242632545](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1465#discussion_r242632545) [](ancestors = 242632545)